numpy/core/src/private/lowlevel_strided_loops.h

```diff
@@ -414,20 +414,23 @@ PyArray_PrepareThreeRawArrayIter(int ndim, npy_intp *shape,
                             char **out_dataC, npy_intp *out_stridesC);
 
 /*
- * Return number of elements that must be peeled from
- * the start of 'addr' with 'nvals' elements of size 'esize'
- * in order to reach 'alignment'.
- * alignment must be a power of two.
- * see npy_blocked_end for an example
+ * Return number of elements that must be peeled from the start of 'addr' with
+ * 'nvals' elements of size 'esize' in order to reach 'alignment'.  alignment
+ * must be a power of two.  see npy_blocked_end for an example
  */
-static NPY_INLINE npy_uintp
+static NPY_INLINE npy_intp
 npy_aligned_block_offset(const void * addr, const npy_uintp esize,
-                         const npy_uintp alignment, const npy_uintp nvals)
+                         const npy_uintp alignment, const npy_intp nvals)
```

Comment:
All of these feel like they ought to be `size_t` to me, and this feels like a step further in the wrong direction

Reply:
[PEP 353](https://www.python.org/dev/peps/pep-0353/#why-not-py-intptr-t) seems to be against [u]intp too

Reply:
I feel just the opposite, unsigned integers are a PITA and should be avoided ;)

Reply:
I suppose what I actually worry about is `ssize_t` vs `intptr_t`. The unsigned-ness is something python agrees with you on.

Reply:
In my experience, code making much use of unsigned integers is invariably buggy and hard to understand; it is hard to get right. For pointer manipulation, which is where we mostly use `npy_intp`,  `intptr_t` seems natural as it is intended to hold pointer differences. I'd like to see just how things would work on a modern segmented architecture,  say IBM AS/400, where full memory access uses 16 byte pointers, and so does `void *`, whereas programs often run in smaller areas where 8 byte pointers are adequate. From the definition of `size_t`, I suspect it would also need to be 16 bytes, but just don't know how things actually work in practice. As far as numpy goes, we could probably simply redefine `npy_intp` if a problem arises. Changing the name would be cleaner, but a total mess downstream ...

Reply:
> intptr_t seems natural as it is intended to hold pointer differences.

Isn't that what `ptrdiff_t` is for?

I'm not sure if `ptrdiff_t` is the same as `ssize_t`, `intptr_t`, or both.

Reply:
Given that the definition of new types is something of a growth industry, I don't think it is a solved problem. I would hope that a compiler on a huge machine might have a compiler flag that would set the memory model so that ordinary imported C code would run reasonably well. That said,  `ptrdif_t` sounds good, but we would need to know what is was in practice. For instance, would it need at least one more bit than `size_t`? OTOH, `Py_ssize_t` would seem to limit array sized to half the maximum, which might matter on tiny segmented architectures. I think this all needs to be addressed on a specific hardware basis when/if it becomes a problem, and then we redefine `npy_intp` to what works best.

Somewhat off topic, I also think  `void *` is overused, too bad `char *` was made to cover the same range. Or does it? We really need to consult a language specialist to understand some of these things ...

Reply:
> might have a compiler flag that would set the memory model so that ordinary imported C code would run reasonably well.

Or maybe folks just run things on virtual machines ...

Reply:
> I think this all needs to be addressed on a specific hardware basis when/if it becomes a problem

I think I agree here.